x-pack/plugins/security_solution/public/detection_engine/rule_management/api/api.ts

```diff
@@ -213,48 +213,45 @@ export interface BulkActionResponse {
   };
 }
 
-export interface BulkActionProps {
-  action: Exclude<BulkAction, BulkAction.export>;
-  query?: string;
-  ids?: string[];
-  edit?: BulkActionEditPayload[];
-  isDryRun?: boolean;
-}
+export type QueryOrIds = { query: string } | { ids: string[] };
```

Comment:
What do you think about using a discriminated union instead?

```ts
export type QueryOrIds = { query: string; ids?: undefined } | { query?: undefined; ids: string[] };
```

Reply:
A discriminated union is discriminated by some field. In this case there is no way to discriminate types so `{ query: string; ids?: undefined } | { query?: undefined; ids: string[] }` is the same as `{ query: string } | { ids: string[] }` which is the same as `{ query?: string; ids: string[]; }` if compare which object type can be assigned.

`string | string[]` looks as a better options here

Reply:
> A discriminated union is discriminated by some field. In this case there is no way to discriminate types so  { query: string; ids?: undefined } | { query?: undefined; ids: string[] } is the same as { query: string } | { ids: string[] } which is the same as { query?: string; ids?: string[]; }

No, this is not correct. Both `query ` and `ids` properties become a common discriminant in this case as they have mutually exclusive types in the union:

![Screenshot 2022-10-28 at 10 48 27](https://user-images.githubusercontent.com/1938181/198545833-cdc21913-8103-49c1-b314-88974e54c77f.png)

Reply:
Hm, I see another picture. I tried it inside Kibana and [TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAigrhATiA8ogkgEwM5QLxQDeUAjgsgFxTbCICWAdgOYDcUdOA-FTfcwNoBdKAF8oAHyKlyIbtVqNW7HDwUDhIlgCgtAMzgMAxsDoB7BlEiJdpxAFsAgsbMMAFGACGiD3exV4SKgYOACUVABuphxEIjpWNvZOJuauxGSBVADk2KZ20OnImQA0yn5Q-JkcAIzFUJWYAEyZGiHa8baOzimEIq06huY0ULoMAApePrgEhFpQ0hl1OXnzhUWzpVQV1bX1TYJrmnFICZ3JbiPj3r59WgMMQx4MpsAAFkgAYmMTvvhE6wUgLJLfIyYrrDhlLaYGolXbNA5QDy4O40bRadqJLpuR7PN6IT6XSY3IA). TS does not complain on passing `{ query: 'myQuery', ids: ['id'] }`.

<img width="673" alt="image" src="https://user-images.githubusercontent.com/3775283/198555071-2f4aeddc-442e-4395-a007-15134389e6d8.png">

Reply:
Sorry, your `QueryOrIds` type is not correct 🙂 You should use this one:

```ts
type QueryOrIds = { query: string; ids?: undefined } | { query?: undefined; ids: string[] };
```

Reply:
Yeah, sorry, it's my bad. I didn't pay enough attention to `undefined`. Basically the key is still assignable but this way it doesn't not create issues.